### PR TITLE
[Parallel Iterator] Local Shuffle

### PR DESCRIPTION
**Pull request by amogkam**
## Why are these changes needed?
Shuffling functionality is needed for the Parallel Iterator library. This PR adds local shuffling methods to Ray ParallelIterator library as well as corresponding tests

## Checks

- [ X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [X ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/experimental/iter.py [hunk 1]
```diff
@@ -238,6 +239,17 @@ def combine(self, fn: Callable[[T], List[U]]) -> "ParallelIterator[U]":
         it.name = self.name + ".combine()"
         return it
 
+    def local_shuffle(self, shuffle_buffer_size: int) -> "ParallelIterator[T]":
+        """Shuffles elements of each shard independently.
```
Comment (ericl) [inline python/ray/experimental/iter.py]: Consider adding an example here similar to for gather_sync().

**Review by ericl — APPROVED** [review]
Please fill out the PR description and run `scripts/format.sh` to lint the files.

File: python/ray/experimental/iter.py [hunk 2]
```diff
@@ -527,6 +539,24 @@ def apply_flatten(it):
             self.local_transforms + [apply_flatten],
             name=self.name + ".flatten()")
 
+    def shuffle(self, shuffle_buffer_size: int) -> "LocalIterator[T]":
```
Comment (richardliaw) [inline python/ray/experimental/iter.py]: Can you please document this?

File: python/ray/experimental/iter.py [hunk 3]
```diff
@@ -1,6 +1,7 @@
 from typing import TypeVar, Generic, Iterable, List, Callable, Any
 
 import ray
+import random
```
Comment (robertnishihara) [inline python/ray/experimental/iter.py]: For imports, please follow the order described in http://google.github.io/styleguide/pyguide.html#313-imports-formatting

*Imports are always put at the top of the file, just after any module comments and docstrings and before module globals and constants. Imports should be grouped from most generic to least generic:*

*Python future import statements. For example:*

    from __future__ import absolute_import
    from __future__ import division
    from __future__ import print_function

*Python standard library imports. For example:*

    import sys

*third-party module or package imports. For example:*

    import tensorflow as tf

*Code repository sub-package imports. For example:*

    from otherproject.ai import mind

File: python/ray/experimental/iter.py (hunk 2 above)
Comment (robertnishihara) [reply python/ray/experimental/iter.py]: docstring style is shown in https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: python/ray/experimental/iter.py [hunk 4]
```diff
@@ -238,6 +239,17 @@ def combine(self, fn: Callable[[T], List[U]]) -> "ParallelIterator[U]":
         it.name = self.name + ".combine()"
         return it
 
+    def local_shuffle(self, shuffle_buffer_size: int) -> "ParallelIterator[T]":
+        """Shuffles elements of each shard independently.
+        """
```
Comment (robertnishihara) [inline python/ray/experimental/iter.py]: Also document the args and returns (if any) in docstring.

File: python/ray/tests/test_iter.py [hunk 5]
```diff
@@ -70,6 +70,38 @@ def test_filter(ray_start_regular_shared):
...
+    assert list(it3.gather_sync()) == list(range(10))
+
+    # statistical test
+    it4 = from_items(
+        [0, 1] * 10000, num_shards=1).local_shuffle(shuffle_buffer_size=100)
+    result = "".join(it4.gather_sync().for_each(str))
+    freq_dict = {}
+    for i in range(0, len(result), 2):
+        s = result[i:i + 2]
+        freq_dict[s] = freq_dict.get(s, 0) + 1
```
Comment (robertnishihara) [inline python/ray/tests/test_iter.py]: The counts can be done a little more cleanly with something like

```python
from collections import Counter

counts = Counter(zip(l[:-1], l[1:]))
```

**Comment by ericl** [comment]
`test_iter` is timing out in travis ^^

File: python/ray/experimental/iter.py [hunk 6]
```diff
@@ -238,6 +239,46 @@ def combine(self, fn: Callable[[T], List[U]]) -> "ParallelIterator[U]":
         it.name = self.name + ".combine()"
         return it
 
+    def local_shuffle(self, shuffle_buffer_size: int,
+                      seed: int = None) -> "ParallelIterator[T]":
+        """Remotely shuffle items of each shard independently
+
+        Arguments:
```
Comment (robertnishihara) [inline python/ray/experimental/iter.py]:
```suggestion
        Args:
```

Could you make this same change everywhere in this file?

File: python/ray/experimental/iter.py [hunk 7]
```diff
@@ -238,6 +239,46 @@ def combine(self, fn: Callable[[T], List[U]]) -> "ParallelIterator[U]":
...
+            shuffle_buffer_size elements and randomly samples elements from
+            this buffer, replacing the selected elements with new elements.
+            For perfect shuffling, this argument should be greater than or
+            equal to the largest iterator size.
+            seed (int): Seed to use for
+            randomness. Default value is None.
+
+        Returns:
+            Returns a ParallelIterator with a local shuffle applied on the
+            base iterator
```
Comment (robertnishihara) [inline python/ray/experimental/iter.py]:
```suggestion
                base iterator
```

The second line should be indented

File: python/ray/experimental/iter.py [hunk 8]
```diff
@@ -238,6 +239,46 @@ def combine(self, fn: Callable[[T], List[U]]) -> "ParallelIterator[U]":
...
         return it
 
+    def local_shuffle(self, shuffle_buffer_size: int,
+                      seed: int = None) -> "ParallelIterator[T]":
+        """Remotely shuffle items of each shard independently
+
+        Arguments:
+            shuffle_buffer_size (int): The algorithm fills a buffer with
+            shuffle_buffer_size elements and randomly samples elements from
+            this buffer, replacing the selected elements with new elements.
```
Comment (robertnishihara) [inline python/ray/experimental/iter.py]:
```suggestion
                this buffer, replacing the selected elements with new elements.
```

Subsequent lines should be indented. Same comment applies in a number of places.

**Comment by amogkam** [comment]
@ericl Some tests are still failing but they are unrelated to what I wrote. `test_iter` passes.

**Pull request merged**